File: backend/agents/simulation-client-agent.ts
/**
 * Simulation Client Agent
 * Simulates realistic client personas in training conversations
 */

import { BaseAgent, ChatMessage, AgentResponse, ToolHandler } from './base-agent.js';
import { AGENT_NAMES } from './agent-config.js';

export interface ClientProfile {
  name: string;
  age: number;
  occupation: string;
  income: string;
  family: string;
  assets?: string[];
  debts?: string[];
  goals?: string[];
  fusionModelTraits?: Record<string, number>;
}

export interface EmotionalState {
  trust: number;
  frustration: number;
  openness: number;
  engagement: number;
  flags?: {
    veryLowTrust?: boolean;
    highFrustration?: boolean;
    lowOpenness?: boolean;
  };
}

export interface ObjectiveProgress {
  rapport: number;
  needs: number;
  objections: number;
  recommendations: number;
  explanation?: string;
}

export interface SimulationContext {
  sessionId: string;
  clientProfile: ClientProfile;
  personalitySettings: {
    mood: string;
    archetype: string;
    traits: Record<string, number>;
    influence: string;
    communicationStyle?: string;
  };
  simulationSettings: {
    industry: string;
    subcategory?: string;
    difficulty: string;
    competencies?: string[];
    focusAreas?: Array<{ id: string; name: string }>;
  };
  emotionalState?: EmotionalState;
  objectives?: ObjectiveProgress;
}

// In-memory context store (in production, use Redis or database)
const contextStore = new Map<string, SimulationContext>();

/**
 * Simulation Client Agent Implementation
 */
export class SimulationClientAgent extends BaseAgent {
  constructor() {
    super(AGENT_NAMES.SIMULATION_CLIENT);
  }

  /**
   * Register tool handlers
   */
  protected registerToolHandlers(): void {
    this.toolHandlers.set('get_client_profile', this.handleGetClientProfile.bind(this));
    this.toolHandlers.set('get_emotional_state', this.handleGetEmotionalState.bind(this));
    this.toolHandlers.set('track_objectives', this.handleTrackObjectives.bind(this));
  }

  /**
   * Tool: Get client profile
   */
  private async handleGetClientProfile(args: { session_id: string }): Promise<ClientProfile | { error: string }> {
    const context = contextStore.get(args.session_id);
    if (!context) {
      return { error: `Session not found: ${args.session_id}` };
    }
    return context.clientProfile;
  }

  /**
   * Tool: Get emotional state
   */
  private async handleGetEmotionalState(args: { session_id: string }): Promise<EmotionalState> {
    const context = contextStore.get(args.session_id);
    if (!context || !context.emotionalState) {
      // Return default emotional state
      return {
        trust: 50,
        frustration: 0,
        openness: 50,
        engagement: 50,
      };
    }
    return context.emotionalState;
  }

  /**
   * Tool: Track objectives
   */
  private async handleTrackObjectives(args: {
    session_id: string;
    rapport: number;
    needs: number;
    objections: number;
    recommendations: number;
    explanation: string;
  }): Promise<{ success: boolean; objectives: ObjectiveProgress }> {
    const context = contextStore.get(args.session_id);
    const objectives: ObjectiveProgress = {
      rapport: args.rapport,
      needs: args.needs,
      objections: args.objections,
      recommendations: args.recommendations,
      explanation: args.explanation,
    };

    if (context) {
      context.objectives = objectives;
      contextStore.set(args.session_id, context);
    }

    return { success: true, objectives };
  }

  /**
   * Set simulation context for a session
   */
  setContext(sessionId: string, context: SimulationContext): void {
    context.sessionId = sessionId;
    contextStore.set(sessionId, context);
  }

  /**
   * Get simulation context for a session
   */
  getContext(sessionId: string): SimulationContext | undefined {
    return contextStore.get(sessionId);
  }

  /**
   * Clear simulation context for a session
   */
  clearContext(sessionId: string): void {
    contextStore.delete(sessionId);
  }

  /**
   * Update emotional state for a session
   */
  updateEmotionalState(sessionId: string, state: Partial<EmotionalState>): void {
    const context = contextStore.get(sessionId);
    if (context) {
      context.emotionalState = {
        ...(context.emotionalState || { trust: 50, frustration: 0, openness: 50, engagement: 50 }),
        ...state,
      };
      contextStore.set(sessionId, context);
    }
  }

  /**
   * Generate client response with context
   */
  async generateResponse(
    messages: ChatMessage[],
    simulationContext: SimulationContext
  ): Promise<AgentResponse & { objectiveProgress?: ObjectiveProgress }> {
    // Store context for tool access
    this.setContext(simulationContext.sessionId, simulationContext);

    // Build enhanced context for the agent
    const enhancedContext = {
      sessionId: simulationContext.sessionId,
      clientProfile: simulationContext.clientProfile,
      personality: simulationContext.personalitySettings,
      industry: simulationContext.simulationSettings.industry,
      subcategory: simulationContext.simulationSettings.subcategory,
      difficulty: simulationContext.simulationSettings.difficulty,
      focusAreas: simulationContext.simulationSettings.focusAreas,
      emotionalState: simulationContext.emotionalState,
      difficultyGuidelines: this.getDifficultyGuidelines(simulationContext.simulationSettings.difficulty),
    };

    // Call base chat method
    const response = await this.chat(messages, enhancedContext);

    // Extract objective progress from tool calls
    let objectiveProgress: ObjectiveProgress | undefined;
    const objectiveCall = response.toolCalls?.find(tc => tc.name === 'track_objectives');
    if (objectiveCall?.result?.objectives) {
      objectiveProgress = objectiveCall.result.objectives;
    } else {
      // Get from context if updated during the run
      const updatedContext = this.getContext(simulationContext.sessionId);
      objectiveProgress = updatedContext?.objectives;
    }

    return {
      ...response,
      objectiveProgress,
    };
  }

  /**
   * Get difficulty guidelines for prompting
   */
  private getDifficultyGuidelines(difficulty: string): string {
    switch (difficulty.toLowerCase()) {
      case 'beginner':
        return `Be friendly, cooperative, and open. Provide information readily when asked. You have basic financial knowledge but need explanations for industry-specific concepts. Share your financial details, family situation, and goals when asked directly.`;
      case 'intermediate':
        return `Be somewhat reserved and hesitant to share all information immediately. Some financial details and goals should only be revealed when asked specifically or when trust is established. You have moderate financial knowledge. Do not volunteer detailed financial information unless specifically asked.`;
      case 'advanced':
        return `Be skeptical, challenging, and resistant initially. Question recommendations, raise objections, and only reveal sensitive information after significant trust-building. You have substantial financial knowledge but may have misconceptions that need correction. Be very guarded with information and require the advisor to demonstrate expertise before opening up.`;
      default:
        return `Be friendly and cooperative, with a balanced approach to sharing information.`;
    }
  }
}

// Export singleton instance
export const simulationClientAgent = new SimulationClientAgent();

export default simulationClientAgent;

File: backend-python/app/agents/simulation_client_agent.py
"""
Simulation Client Agent
Simulates client personas in training conversations
"""

import json
import logging
from typing import Any

from azure.ai.agents.models import ToolDefinition, FunctionTool, FunctionToolDefinition, FunctionDefinition

from app.agents.base_agent import BaseAgent

logger = logging.getLogger(__name__)


SIMULATION_CLIENT_INSTRUCTIONS = """You are an AI client in a training simulation. Your role is to behave as a realistic client with the personality traits, background, and needs specified in your profile.

IMPORTANT GUIDELINES:
1. You are the CLIENT, not the advisor. Respond as if you are seeking financial advice.
2. Stay in character throughout the conversation.
3. Respond naturally and conversationally, avoiding robotic language.
4. Never reference that you are an AI or break character.
5. Use the available tools to:
   - Get your client profile details when needed
   - Understand your emotional state for appropriate responses
   - Track conversation objectives progress

DIFFICULTY BEHAVIORS:
- Beginner: Be friendly, cooperative, and open. Provide information readily.
- Intermediate: Be somewhat reserved. Only reveal details when asked specifically or when trust is established.
- Advanced: Be skeptical and challenging. Question recommendations and raise objections.

For your first response, introduce yourself briefly with just your name and a general reason for meeting with the advisor."""


def get_difficulty_guidelines(difficulty: str) -> str:
    """Get difficulty-specific guidelines."""
    difficulty_lower = difficulty.lower()
    if difficulty_lower == "beginner":
        return "Be friendly, cooperative, and open. Provide information readily when asked."
    elif difficulty_lower == "intermediate":
        return "Be somewhat reserved. Only reveal details when asked specifically or when trust is established."
    elif difficulty_lower == "advanced":
        return "Be skeptical, challenging, and resistant. Question recommendations and raise objections."
    return "Be friendly and cooperative, with a balanced approach."


class SimulationClientAgent(BaseAgent):
    """
    Agent that simulates client personas in training conversations.
    Used for generating realistic client responses in advisor simulations.
    """

    def __init__(self):
        super().__init__(
            name="simulation-client",
            instructions=SIMULATION_CLIENT_INSTRUCTIONS,
            model=None,  # Use default model from settings
        )
        self._client_profile: dict[str, Any] = {}
        self._personality_settings: dict[str, Any] = {}
        self._simulation_settings: dict[str, Any] = {}

    def get_tools(self) -> list[ToolDefinition]:
        """Get tool definitions for the simulation client agent."""
        return [
            FunctionToolDefinition(
                function=FunctionDefinition(
                    name="get_client_profile",
                    description="Get the current client profile including name, age, occupation, income, goals, and background.",
                    parameters={
                        "type": "object",
                        "properties": {},
                        "required": [],
                    },
                ),
            ),
            FunctionToolDefinition(
                function=FunctionDefinition(
                    name="get_emotional_state",
                    description="Get the client's current emotional state and mood for appropriate response tone.",
                    parameters={
                        "type": "object",
                        "properties": {},
                        "required": [],
                    },
                ),
            ),
            FunctionToolDefinition(
                function=FunctionDefinition(
                    name="track_objectives",
                    description="Track and evaluate progress on conversation objectives like rapport building and needs assessment.",
                    parameters={
                        "type": "object",
                        "properties": {
                            "rapport_progress": {
                                "type": "number",
                                "description": "Progress on building rapport (0-100)",
                            },
                            "needs_progress": {
                                "type": "number",
                                "description": "Progress on needs assessment (0-100)",
                            },
                            "observation": {
                                "type": "string",
                                "description": "Brief observation about advisor's approach",
                            },
                        },
                        "required": [],
                    },
                ),
            ),
        ]

    async def handle_tool_call(self, tool_name: str, arguments: dict[str, Any]) -> str:
        """Handle tool calls from the agent."""
        context = getattr(self, "_current_context", {})

        if tool_name == "get_client_profile":
            profile = context.get("client_profile", self._client_profile)
            return json.dumps({
                "name": profile.get("name", "Unknown Client"),
                "age": profile.get("age", "Unknown"),
                "occupation": profile.get("occupation", "Unknown"),
                "income": profile.get("income", "Unknown"),
                "family": profile.get("family", "Unknown"),
                "goals": profile.get("goals", []),
                "concerns": profile.get("concerns", []),
                "background": profile.get("background", ""),
            })

        elif tool_name == "get_emotional_state":
            personality = context.get("personality_settings", self._personality_settings)
            simulation = context.get("simulation_settings", self._simulation_settings)
            return json.dumps({
                "mood": personality.get("mood", "neutral"),
                "archetype": personality.get("archetype", "Standard Client"),
                "difficulty": simulation.get("difficulty", "beginner"),
                "traits": personality.get("traits", {}),
                "guidelines": get_difficulty_guidelines(simulation.get("difficulty", "beginner")),
            })

        elif tool_name == "track_objectives":
            # This is primarily for the agent to internally track, but we log it
            logger.debug(f"[SimulationClientAgent] Objectives tracked: {arguments}")
            return json.dumps({
                "status": "tracked",
                "rapport": arguments.get("rapport_progress", 0),
                "needs": arguments.get("needs_progress", 0),
            })

        return json.dumps({"error": f"Unknown tool: {tool_name}"})

    async def generate_response(
        self,
        messages: list[dict[str, Any]],
        client_profile: dict[str, Any],
        personality_settings: dict[str, Any] | None = None,
        simulation_settings: dict[str, Any] | None = None,
        session_id: str | None = None,
    ) -> dict[str, Any]:
        """
        Generate a client response for the simulation.

        Args:
            messages: Conversation history
            client_profile: Client profile data
            personality_settings: Optional personality settings
            simulation_settings: Optional simulation settings
            session_id: Optional session ID for thread management

        Returns:
            Response dictionary with message and metadata
        """
        if not self.is_initialized:
            raise RuntimeError("SimulationClientAgent not initialized")

        # Store context for tool calls
        self._client_profile = client_profile
        self._personality_settings = personality_settings or {}
        self._simulation_settings = simulation_settings or {}

        # Build context for tools
        context = {
            "client_profile": client_profile,
            "personality_settings": personality_settings or {},
            "simulation_settings": simulation_settings or {},
        }

        # Build enhanced prompt with profile context
        profile_context = self._build_profile_context(
            client_profile,
            personality_settings or {},
            simulation_settings or {},
        )

        # Get or create thread
        thread_id = await self.get_or_create_thread(session_id or "default")

        # Format conversation for agent
        conversation_text = self._format_conversation(messages)
        prompt = f"{profile_context}\n\nConversation so far:\n{conversation_text}\n\nRespond as the client:"

        # Send message and get response
        response_text = await self.send_message(thread_id, prompt, context)

        return {
            "message": response_text,
            "source": "azure-agent",
            "agent_id": self.agent_id,
        }

    def _build_profile_context(
        self,
        client_profile: dict[str, Any],
        personality_settings: dict[str, Any],
        simulation_settings: dict[str, Any],
    ) -> str:
        """Build profile context for the prompt."""
        difficulty = simulation_settings.get("difficulty", "beginner")
        guidelines = get_difficulty_guidelines(difficulty)

        return f"""Current Client Context:
- Name: {client_profile.get('name', 'Unknown')}
- Age: {client_profile.get('age', 'Unknown')}
- Occupation: {client_profile.get('occupation', 'Unknown')}
- Income: {client_profile.get('income', 'Unknown')}
- Family Status: {client_profile.get('family', 'Unknown')}
- Goals: {', '.join(client_profile.get('goals', [])) if client_profile.get('goals') else 'Not specified'}

Personality:
- Mood: {personality_settings.get('mood', 'neutral')}
- Archetype: {personality_settings.get('archetype', 'Standard Client')}

Industry: {simulation_settings.get('industry', 'Unknown')}{f" - {simulation_settings.get('subcategory')}" if simulation_settings.get('subcategory') else ''}
Difficulty: {difficulty}

Behavior Guidelines: {guidelines}"""

    def _format_conversation(self, messages: list[dict[str, Any]]) -> str:
        """Format conversation history for the prompt."""
        formatted = []
        for msg in messages:
            role = msg.get("role", "unknown")
            content = msg.get("content", "")
            if role == "user":
                formatted.append(f"Advisor: {content}")
            elif role == "assistant":
                formatted.append(f"Client: {content}")
        return "\n".join(formatted) if formatted else "(Start of conversation)"
